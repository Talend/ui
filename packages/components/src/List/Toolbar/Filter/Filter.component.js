import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';
import DebounceInput from 'react-debounce-input';
import FormControl from 'react-bootstrap/lib/FormControl';
import get from 'lodash/get';
import keycode from 'keycode';

import defaultTranslateFn from '../../../translate';
import { Action } from '../../../Actions';
import Icon from '../../../Icon';
import theme from './Filter.scss';

function onKeyDown(event, escAction, enterAction) {
	switch (event.keyCode) {
	case keycode.codes.enter:
		if (enterAction) {
			enterAction(event);
		}
		break;
	case keycode.codes.esc:
		if (escAction) {
			escAction(event);
		}
		break;
	default:
		break;
	}
}

function FilterInput(props) {
	const {
		id,
		debounceMinLength,
		debounceTimeout,
		onBlur,
		onFocus,
		onFilter,
		onToggle,
		placeholder,
		value,
	} = props;

	const inputProps = {
		id,
		name: 'search',
		type: 'search',
		value,
		placeholder,
		className: theme.search,
		'aria-label': 'Filter',
		onBlur: onBlur && (event => onBlur(event, event.target.value)),
		onFocus: onFocus && (event => onFocus(event, event.target.value)),
		onChange: event => onFilter(event, event.target.value),
		onKeyDown: event => onKeyDown(event, onToggle, onBlur),
		autoFocus: true,
	};

	if (debounceMinLength || debounceTimeout) {
		return (<DebounceInput
			{...inputProps}
			element={FormControl}
			minLength={debounceMinLength}
			debounceTimeout={debounceTimeout}
		/>);
	}
	return (<FormControl {...inputProps} />);
}

FilterInput.propTypes = {
	id: PropTypes.string,
	debounceMinLength: PropTypes.number,
	debounceTimeout: PropTypes.number,
	onBlur: PropTypes.func,
	onFocus: PropTypes.func,
	onFilter: PropTypes.func.isRequired,
	onToggle: PropTypes.func,
	placeholder: PropTypes.string,
	value: PropTypes.string,
};

/**
 * @param {object} props react props
 * @example
 <Filter id="my-filter" docked="false" onFilter="filter()"></Filter>
 */
function Filter(props) {
	const {
		id,
		debounceMinLength,
		debounceTimeout,
		docked,
		highlight,
		onBlur,
		onFocus,
		onFilter,
		onToggle,
		placeholder,
		value,
		t,
	} = props;
	if (docked) {
		return (
			<Action
				id={id}
				className="navbar-right"
				onClick={onToggle}
				label={t('LIST_FILTER_TOGGLE', { defaultValue: 'Toggle filter' })}
				hideLabel
				icon="talend-search"
				bsStyle="link"
			/>
		);
	}

	function onSubmit(event) {
		event.preventDefault();
		return onFilter(event, get(event, 'target.search.value'));
	}

	const classes = classNames(
		'navbar-form',
		'navbar-right',
		theme.filter,
		{ [theme.highlight]: highlight },
	);

	return (
		<form
			className={classes}
			role="search"
			onSubmit={onSubmit}
		>
			<Icon name="talend-search" className={theme['search-icon']} />
			<div className="form-group">
				<FilterInput
					id={id && `${id}-input`}
					debounceMinLength={debounceMinLength}
					debounceTimeout={debounceTimeout}
					onBlur={onBlur}
					onFocus={onFocus}
					onFilter={onFilter}
					onToggle={onToggle}
					placeholder={placeholder}
					value={value}
				/>
				<Action
					id={id && `${id}-cross-icon`}
					bsStyle="link"
					icon="talend-cross"
					label={t('LIST_FILTER_REMOVE', { defaultValue: 'Remove filter' })}
					hideLabel
					onClick={onToggle}
				/>
			</div>
		</form>
	);
}

Filter.propTypes = {
	id: PropTypes.string,
	debounceMinLength: PropTypes.number,
	debounceTimeout: PropTypes.number,
	docked: PropTypes.bool,
	onBlur: PropTypes.func,
	onFocus: PropTypes.func,
	onFilter: PropTypes.func.isRequired,
	onToggle: PropTypes.func.isRequired,
	highlight: PropTypes.bool,
	placeholder: PropTypes.string,
	value: PropTypes.string,
	t: PropTypes.func.isRequired,
};

Filter.defaultProps = {
	docked: true,
	placeholder: 'Filter',
	t: defaultTranslateFn,
};

export default Filter;
