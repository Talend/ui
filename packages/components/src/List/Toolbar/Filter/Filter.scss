// Filter styles
$tc-list-filter-input-size: 200px !default;
$tc-list-filter-highlight-color: #F6E5BB !default;

.filter {
	display: flex;
	align-items: center;
	padding: 0;
	background: $white;

	.search-icon {
		margin: $padding-small;
	}

	:global(.form-group.form-group) {
		padding: 0;
		background: none;
	}

	.search {
		width: $tc-list-filter-input-size;
		animation: reveal 1s cubic-bezier(0.4, 0, 0.2, 1);

		&::-ms-clear { // sass-lint:disable-line no-vendor-prefixes
			display: none;
		}
	}
}

.highlight {
	background-color: $tc-list-filter-highlight-color;
}

@keyframes reveal {
	0% {
		opacity: 0;
		width: 0;
	}

	50% {
		opacity: 1;
	}

	100% {
		width: $tc-list-filter-input-size;
	}
}
