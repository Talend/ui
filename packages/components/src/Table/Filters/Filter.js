/**
 * This class represents a filter for the elements of a table.
 * A filter is identified by a unique string identifier.
 * A filter can be activated or deactivated.
 */
export default class Filter {
	/**
	 * @param {string} id - filter unique identifier
	 * @param {string} key - the key used to get values from data accessor
	 * @param {boolean} active - Initial state of the filter
	 */
	constructor(id, key, active) {
		this.id = id;
		this.key = key;
		this.active = active;
	}

	getId() {
		return this.id;
	}

	getKey() {
		return this.key;
	}

	/**
	 * Set the filter active or inactive.
	 * @param {boolean} active
	 */
	setActive(active) {
		this.active = active;
	}

	/**
	 * Indicates if the filter is active or not.
	 */
	isActive() {
		return this.active;
	}

	/**
	 * This methods indicates if an element of a list is visible or not.
	 * When returns true, the element is visible (not filtered).
	 * @param {object} dataAccessor - element data getter
	 * @param {object} element - an element of the table
	 */
	select(dataAccessor, element) {
		if (this.isActive()) {
			const value = dataAccessor.getRowData(element, this.key);
			return this.selectValue(value);
		}
		return true;
	}

	// This method is intended to be subclassed.
	/* eslint-disable class-methods-use-this */
	selectValue(value) {
		return Boolean(value);
	}
	/* eslint-enable class-methods-use-this */
}
