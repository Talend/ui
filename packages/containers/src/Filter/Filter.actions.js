import { actions } from '@talend/react-cmf';

export function saveFilterToStore(event, { componentName, key, collectionFiltered }) {
	return actions.componentsActions.mergeState(componentName, key, {
		filterInputValue: event.target.value,
		collectionFiltered,
	});
}

export function updateCollectionFilterToStore(event, { componentName, key, collectionFiltered }) {
	return actions.componentsActions.mergeState(componentName, key, {
		collectionFiltered,
	});
}
