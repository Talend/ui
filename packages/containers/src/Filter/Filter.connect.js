import { Map, List } from 'immutable';
import { api, cmfConnect } from '@talend/react-cmf';
import Container from './Filter.container';
import { getFilterInputValue } from './Filter.selectors';

const DEFAULT_STATE = Map({ filterInputValue: '', collectionFiltered: List() });

function mapStateToProps(state, ownProps) {
	if (!ownProps.actionId && !ownProps.name) {
		return {};
	}
	const actionInfo = api.action.getActionInfo(
		{
			registry: api.registry.getRegistry(),
			store: {
				getState: () => state,
			},
		},
		ownProps.actionId || ownProps.getname,
	);
	return {
		collectionToFilter: state.cmf.collections.get(actionInfo['collection-to-filter']),
		filterInputValue: getFilterInputValue(state, actionInfo.id),
		actionInfo,
	};
}

export default cmfConnect({
	componentId: ownProps => (ownProps && ownProps.id) || 'Filter',
	defaultState: DEFAULT_STATE,
	mapStateToProps,
})(Container);
