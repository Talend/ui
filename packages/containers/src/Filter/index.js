import Filter from './Filter.component';

export default Filter;
