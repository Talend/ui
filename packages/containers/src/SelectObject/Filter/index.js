import Filter from './Filter.connect';

export default Filter;
