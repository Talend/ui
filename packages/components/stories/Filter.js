import React from 'react';
import {
	// eslint-disable-line import/no-extraneous-dependencies
	storiesOf,
	action,
} from '@storybook/react';

import { Filter, IconsProvider } from '../src/';

const propsDockToggle = {
	id: 'FILTER-dockAndToggle',
	docked: false,
	navbar: true,
	toggeable: true,
	onFilter: action('onFilter'),
	onBlur: action('onBlur'),
	onFocus: action('onFocus'),
	onToggle: action('onToggle'),
	placeholder: 'My placeholder',
	highlight: false,
};

const propsNoDockToggle = {
	id: 'FILTER-noDockAndNoToggle',
	docked: false,
	navbar: false,
	toggeable: false,
	onFilter: action('onFilter'),
	onBlur: action('onBlur'),
	onFocus: action('onFocus'),
	onToggle: action('onToggle'),
	placeholder: 'My placeholder',
	highlight: false,
};

const stories = storiesOf('Filter', module);
if (!stories.addWithInfo) {
	stories.addWithInfo = stories.add;
}

const divStyle = {
	width: '30rem',
};

stories
	.addWithInfo('default-dock and toggle', () => (
		<nav>
			<div style={divStyle}>
				<IconsProvider />
				<Filter {...propsDockToggle} />
			</div>
		</nav>
	))
	.addWithInfo('custom-undock no toggle', () => (
		<div>
			<IconsProvider />
			<Filter {...propsNoDockToggle} />
		</div>
	));
