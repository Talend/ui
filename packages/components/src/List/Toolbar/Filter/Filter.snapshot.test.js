import React from 'react';
import { shallow } from 'enzyme';

import Filter from './Filter.component';

jest.mock('react-dom');

describe('Filter', () => {
	const noOp = jest.fn();

	it('should render', () => {
		// when
		const wrapper = shallow(<Filter onFilter={noOp} onToggle={noOp} />);

		// then
		expect(wrapper.node).toMatchSnapshot();
	});

	it('should render id if provided', () => {
		// when
		const wrapper = shallow(<Filter id="toolbar-filter" onFilter={noOp} onToggle={noOp} />);

		// then
		expect(wrapper.node).toMatchSnapshot();
	});

	it('should render highlighted filter', () => {
		// when
		const wrapper = shallow(<Filter highlight onFilter={noOp} onToggle={noOp} />);

		// then
		expect(wrapper.node).toMatchSnapshot();
	});

	it('should render only toggle icon', () => {
		// when
		const wrapper = shallow(<Filter onFilter={noOp} onToggle={noOp} docked />);

		// then
		expect(wrapper.node).toMatchSnapshot();
	});

	it('should render filter input with given placeholder', () => {
		// when
		const defaultProps = {
			docked: false,
			placeholder: 'find something',
		};
		const wrapper = shallow(<Filter {...defaultProps} />);

		// then
		expect(wrapper.node).toMatchSnapshot();
	});

	it('should render filter input with default placeholder', () => {
		// when
		const defaultProps = {
			docked: false,
		};
		const wrapper = shallow(<Filter {...defaultProps} />);

		// then
		expect(wrapper.node).toMatchSnapshot();
	});
});
