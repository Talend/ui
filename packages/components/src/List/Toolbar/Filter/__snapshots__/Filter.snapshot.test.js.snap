// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter should render 1`] = `
<button
  className="navbar-right btn btn-link"
  disabled={false}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseOver={[Function]}
  role={null}
  type="button"
>
  <svg
    aria-hidden="true"
    className="tc-svg-icon"
    focusable="false"
    title={null}
  >
    <use
      xlinkHref="#talend-search"
    />
  </svg>
</button>
`;

exports[`Filter should render filter input with default placeholder 1`] = `
<form
  className="navbar-form navbar-right"
  onSubmit={[Function]}
  role="search"
>
  <svg
    aria-hidden="true"
    className="tc-svg-icon"
    focusable="false"
    title={null}
  >
    <use
      xlinkHref="#talend-search"
    />
  </svg>
  <div
    className="form-group"
  >
    <input
      aria-label="Filter"
      autoFocus={true}
      className="form-control"
      id={undefined}
      name="search"
      onBlur={undefined}
      onChange={[Function]}
      onFocus={undefined}
      onKeyDown={[Function]}
      placeholder="Filter"
      type="search"
      value={undefined}
    />
    <button
      className="btn btn-link"
      disabled={false}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseOver={[Function]}
      role={null}
      type="button"
    >
      <svg
        aria-hidden="true"
        className="tc-svg-icon"
        focusable="false"
        title={null}
      >
        <use
          xlinkHref="#talend-cross"
        />
      </svg>
    </button>
  </div>
</form>
`;

exports[`Filter should render filter input with given placeholder 1`] = `
<form
  className="navbar-form navbar-right"
  onSubmit={[Function]}
  role="search"
>
  <svg
    aria-hidden="true"
    className="tc-svg-icon"
    focusable="false"
    title={null}
  >
    <use
      xlinkHref="#talend-search"
    />
  </svg>
  <div
    className="form-group"
  >
    <input
      aria-label="Filter"
      autoFocus={true}
      className="form-control"
      id={undefined}
      name="search"
      onBlur={undefined}
      onChange={[Function]}
      onFocus={undefined}
      onKeyDown={[Function]}
      placeholder="find something"
      type="search"
      value={undefined}
    />
    <button
      className="btn btn-link"
      disabled={false}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseOver={[Function]}
      role={null}
      type="button"
    >
      <svg
        aria-hidden="true"
        className="tc-svg-icon"
        focusable="false"
        title={null}
      >
        <use
          xlinkHref="#talend-cross"
        />
      </svg>
    </button>
  </div>
</form>
`;

exports[`Filter should render highlighted filter 1`] = `
<button
  className="navbar-right btn btn-link"
  disabled={false}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseOver={[Function]}
  role={null}
  type="button"
>
  <svg
    aria-hidden="true"
    className="tc-svg-icon"
    focusable="false"
    title={null}
  >
    <use
      xlinkHref="#talend-search"
    />
  </svg>
</button>
`;

exports[`Filter should render id if provided 1`] = `
<button
  className="navbar-right btn btn-link"
  disabled={false}
  id="toolbar-filter"
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseOver={[Function]}
  role={null}
  type="button"
>
  <svg
    aria-hidden="true"
    className="tc-svg-icon"
    focusable="false"
    title={null}
  >
    <use
      xlinkHref="#talend-search"
    />
  </svg>
</button>
`;

exports[`Filter should render only toggle icon 1`] = `
<button
  className="navbar-right btn btn-link"
  disabled={false}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseOver={[Function]}
  role={null}
  type="button"
>
  <svg
    aria-hidden="true"
    className="tc-svg-icon"
    focusable="false"
    title={null}
  >
    <use
      xlinkHref="#talend-search"
    />
  </svg>
</button>
`;
